feat(ROK-696): update package collection while push rpm to koji

qgong opened the pull request:
Contribute the step of updating package collection into pipeline push-rpm-to-koji

## Describe your changes

## Relevant Jira
[ROK-696](https://issues.redhat.com/browse/ROK-696)

## Checklist before requesting a review
- [ ] I have marked as draft or added `do not merge` label if there's a dependency PR
  - If you want reviews on your draft PR, you can add reviewers or add the `release-service-maintainers` handle if you are unsure who to tag
- [ ] My commit message includes `Signed-off-by: My name <email>`
- [ ] I read CONTRIBUTING.MD and [commit formatting](CONTRIBUTING.md#commit-message-formatting-and-standards)
- [ ] I have run the README.md generator script in `.github/scripts/readme_generator.sh` and verified the results using `.github/scripts/check_readme.sh`

seanconroy2021 commented on stepactions/update-package-collection/update-package-collection.yaml:
```diff
@@ -0,0 +1,129 @@
+apiVersion: tekton.dev/v1
+kind: StepAction
```
Could this not be a task? My only concern is that we don't have unit tests for step actions.

xiangge replied:
Please update it as task and put it into push-rpm-to-koji pipeline.

jinqi7 commented on tasks/managed/update-package-collection/update-package-collection.yaml:
```diff
@@ -0,0 +1,255 @@
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: update-package-collection
```
Does the task need to add any unit test?

qgong replied:
> Does the task need to add any unit test?
I am preparing the unit test.

Commit:
feat(ROK-696): add the task of updating package collection

Contribute the Tekton tasks of generating package collection
into the release-service-catalog repo
- After task push rpm build to koji
- After task promote koji draft build

Signed-off-by: Rony Gong <qgong@redhat.com>

xiangge commented on pipelines/managed/push-rpm-to-koji/push-rpm-to-koji.yaml:
```diff
@@ -248,7 +248,7 @@ spec:
           value: $(tasks.collect-task-params.results.extractedValues[2])
         - name: pipelineImage
-          value: $(tasks.collect-task-params.results.extractedValues[3])
+          value: "quay.io/redhat-user-workloads/rhel-on-konflux-tenant/koji-import-utils:on-pr-e0ad44190787415d5c9126fe1cfe1e9576965a65"
```
This shouldn't be here. We need to use the original $(tasks.collect-task-params.results.extractedValues[3])

qgong replied:
+1, I will update it once the new image uploaded

qgong commented:
@jinqi7 @seanconroy2021 Could you help review this patch and trigger the test? Thank you!

jinqi7 commented:
/retest

jinqi7 commented on tasks/managed/update-package-collection-after-promote/update-package-collection-after-promote.yaml:
What's the secret "package-collection-secret" for? Why it's commented and left in the file?

qgong replied:
This secret is still not ready yet, the releng team is still helping add this secret into the releng tenant

seanconroy2021 commented on tasks/managed/update-package-collection/README.md:
```diff
@@ -0,0 +1,23 @@
+# update-package-collection-after-push
```
I feel the task name is very long. Could we shorten it to maybe something like:
- update-package-collection-after-push
- update-package-collection-after-promote

qgong replied:
@seanconroy2021  I am confused for your suggestion here as they are same length of my currecnt name

seanconroy2021 replied:
Sorry, I was meant to remove the 'after' part.  Up to you, just a suggestion :)

qgong replied:
yeap, I am trying to merge this 2 tasks into one if possible

seanconroy2021 commented on tasks/managed/update-package-collection-after-promote/tests/test-promote-koji-draft-build.yaml:
```diff
@@ -0,0 +1,380 @@
+              test_cmd_count 3 uv run pc-manager package-collection add-builds
+
+              echo "All tests completed with exit code: $exit_code"
+              exit $exit_code
```
nit - missing new line

qgong replied:
Ack!

seanconroy2021 replied:
Still present :)

seanconroy2021 commented on tasks/managed/update-package-collection-after-promote/update-package-collection-after-promote.yaml:
```diff
@@ -0,0 +1,490 @@
+        - name: GITLAB_HOST
+          value: "gitlab.cee.redhat.com"
+        - name: GIT_AUTHOR_NAME
+          value: "pkg_coll_auto_bot"
```
Where are these parameters going to come from: the secret or defining them in the data? As this can't be really hardcoded at the end

qgong replied:
After the secret 'package-collection-secret' being uploaded by releng, I will update these code.

xiangge commented on tasks/managed/update-package-collection-after-promote/update-package-collection-after-promote.yaml:
```diff
@@ -0,0 +1,490 @@
+        }
+
+        # Get build information for a component
+        get_component_build_info() {
```
Can we get the component information similar as the draft-push task? Fetching data from snapshot? See the promote-koji-draft-build.yaml 

The reason is once we won't use koji to store the RPM and only use quay.io, this logic won't change much. 

I hope here we won't use any of the koji commands

qgong replied:
The components in this snapshot doesn't mean the components will be 100% push to brew well. My task of update-package-collection-after-promote will verify the builds promote or not in brew, then update it's package collection. IMO, we may consider the promote process is success or not.

And we need fetch the real tags from brew after the promote process.  We can't fetch and use the tags from DATA_FILE and SNAPSHOT_SPEC_FILE

xiangge commented on tasks/managed/update-package-collection-after-push/update-package-collection-after-push.yaml:
```diff
@@ -0,0 +1,475 @@
+
+        # Setup environment variables
+        setup_environment() {
+          export CI_SERVER_HOST="$GITLAB_HOST"
```
You didn't export this on the promote task.

qgong replied:
Ack!

xiangge commented on tasks/managed/update-package-collection-after-push/update-package-collection-after-push.yaml:
```diff
@@ -0,0 +1,475 @@
+          export KEYTAB_FILE=$(jq -r '.pushOptions.pushKeytab.name' "${DATA_FILE}")
+          export KOJI_PROFILE=$(jq -r '.pushOptions.koji_profile' "${DATA_FILE}")
+          export KOJI_TAGS=$(jq -r '.pushOptions.koji_tags // [] | join(" ")' "${DATA_FILE}")
+          export KOJI_IMPORT_DRAFT=$(jq -r '.pushOptions.koji_import_draft' "${DATA_FILE}")
```
```suggestion
          export KOJI_TAGS=$(jq -r '.pushOptions.koji_tags // [] | join(" ")' "${DATA_FILE}")
          export KOJI_IMPORT_DRAFT=$(jq -r '.pushOptions.koji_import_draft' "${DATA_FILE}")
```
Get the promote value inside the pipeline and based the pipeline to set or not set these two values
       - name: promoteBuild
          value: $(tasks.collect-task-params.results.extractedValues[3])
          
 Add the 
          # Extract username from principal
          export USER_NAME=$(echo "$PRINCIPAL" | cut -d'@' -f1)
           # Extract username from principal
           export USER_NAME=$(echo "$PRINCIPAL" | cut -d'@' -f1)
          
 when the promote is true.
 
Plus another comment I left, I feel like there is no much diff between these two tasks. I feel like we can merge them to one and we call two times with different value of the promoteBuild in the push-rpm-to-koji pipeline, what do you think?

qgong reviewed (commented):
replied

xiangge commented on tasks/managed/update-package-collection/update-package-collection.yaml:
```diff
@@ -0,0 +1,496 @@
+                      fi
+                  fi
+              done <<< "$collections_json"
+          else
+              echo "INFO: No package collections found for $comp_name" >&2
+          fi
```
Seems, this is not what I suggested.

I said we only check the is_sidetag for koji_target, but not the collections_json from component name.

Do I misunderstand your logic?

qgong commented:
@jinqi7 @seanconroy2021 @xiangge please help review it, thank you!

seanconroy2021 commented on tasks/managed/update-package-collection/update-package-collection.yaml:
```diff
@@ -0,0 +1,548 @@
+    - name: sourceDataArtifact
+      type: string
+      description: "Produced trusted data artifact"
+    - name: pkgCollectionData
```
You define pkgCollectionData, but you never write to it?

qgong replied:
fixed!

seanconroy2021 commented on tasks/managed/update-package-collection/update-package-collection.yaml:
```diff
@@ -0,0 +1,548 @@
+
+          # Combine koji_target and collections_json
+          # Output koji_target as part of the collections (to stdout)
+          echo "$koji_target"
```
Should this not be koji_tag from looking at line 363. Otherwise why do we set it koji_tag=${koji_target}

seanconroy2021 commented on tasks/managed/update-package-collection/update-package-collection.yaml:
```diff
@@ -0,0 +1,548 @@
+        - name: workDir
+          value: $(params.dataDir)
+        - name: sourceDataArtifact
+          value: $(results.sourceDataArtifact.path)
```
nit - needs a new line

qgong replied:
fixed

seanconroy2021 reviewed (commented):
It might be worth adding few tests  a lot of your  jq  dont have error handling so it wont fail gracefully. What happens if no SRPM file  is present. Then you have trasation error have tested that the retry work even if its mocked.

xiangge commented on tasks/managed/update-package-collection/update-package-collection.yaml:
```diff
@@ -0,0 +1,546 @@
+        }
+
+        # Setup Kerberos authentication
+        setup_kerberos_auth() {
```
```suggestion
        setup_kerberos_auth() {
          # Skip the init as it's not draft push
          if [[ "$KOJI_IMPORT_DRAFT" != "true" ]]; then
            return 0
          fi
```

qgong commented:
> It might be worth adding few tests a lot of your jq dont have error handling so it wont fail gracefully. What happens if no SRPM file is present. Then you have trasation error have tested that the retry work even if its mocked.

@seanconroy2021 I should I add more tests to which part?  or should I add more error checkpoints to avoid execption like:
# Get the most recently modified .src.rpm file
SRPM=$(ls -t ./*.src.rpm 2>/dev/null | head -n1)

if [ -z "$SRPM" ]; then
    echo "Error: No .src.rpm files found" >&2
    exit 1
fi

PACKAGE_NVR=$(basename "$SRPM" .src.rpm)

seanconroy2021 commented on tasks/managed/update-package-collection/update-package-collection.yaml:
```diff
@@ -0,0 +1,586 @@
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: update-package-collection
```
Missing annotations:
```
  annotations:
    tekton.dev/pipelines.minVersion: "0.12.1"
    tekton.dev/tags: release
```

seanconroy2021 reviewed (approved):
LGTM
